### ISSUE TITLE
Abort "updater --no-interaction" if any enabled apps do not have viable update canadidates for the new Server version

### ISSUE THREAD

**Issue by brad2014** [nextcloud/updater#399]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
1.  Update to Nextcloud Hub II 23.0.0 using `sudo -u apache php updater/updater.phar --no-interaction`
2.  Try to log in as IMAP user, watch it fail.

### Expected behaviour
- IMAP users should continue to be able to login after the upgrade reports successful completion, or
- The upgrade should atomically fail if enabled apps are not compatible with it.  (possibly the user can be given the choice to upgrade with incompatible apps disabled in interactive mode, but that should not be the default) 
- A server release, present on the stable channel, should generally be compatible with basic apps (at least those used for login).

### Actual behaviour
- The upgrade claims to succeed.
- I am unable to login as IMAP user (everyone is locked out).
- Several of my installed apps (external_user, talk, totp, calendar) are disabled, because they are not marked as compatible with the stable server version.

### Notes
In the error log, I see
```
  Error     core                      User backend OC_User_IMAP not found.                                                                                                                2021-12-02T14:37:13+00:00  
```
It turns out user_external app was disabled, because it was not compatible with the upgrade version.

### Workaround
Log in as admin, enable all the disabled apps, overriding the "untested" warning.

### Server configuration

**Operating system:**
Fedora 33

**Web server:**
Apache

**Database:**
MySql 10

**PHP version:**
7.4.26

**Nextcloud version:** (see Nextcloud admin page)
23.0.0

**Updated from an older Nextcloud/ownCloud or fresh install:**
updated using stable channel
**Where did you install Nextcloud from:** nextcloud.org

**Are you using external storage, if yes which one:** no
**Are you using encryption:** no
**Are you using an external user-backend, if yes which one:** IMAP

**Comment by joshtrichards** [comment]
I understand your preference to have `--no-interaction` automatically fail if any apps lack compatible update candidates with the new Server version. I'm not sure we can realistically change the default behavior for `--no-interaction` at this point (and it's basically a "don't ask / assume yes" option) since it would be a breaking change for others. However, it may be reasonable to add an additional option that makes the Updater/upgrader behave that way...

> Ah, I see a few hours later some of the apps are updating.

This is the most mysterious part to me.

Not sure why it would be a few hours later. The upgrade isn't complete if there are any existing apps with compatible candidates available to update to. Even if new app versions were published the same day they wouldn't auto install on their own once the upgrader finished.

I can think of a couple possibilities that kind of / sort of / but still not entirely make sense to me (either individually or in combination) to explain what you experienced:

* PHP caching (we recommend restarting web/mod_php or fpm after upgrades if your environment is prone to this / or in general to avoid problems)
* the `occ upgrade` didn't actual run when the `updater.phar` finished its own steps
*  the command `occ upgrade` was manually ran at some point during that couple hour period after the `updater.phar` finished (particularly if `--no-upgrade` was also specified to `updater.phar`)
* you swapped release channels to `beta` to get offered the *then* just released Server v23.0.0 during it's soft release period (where it's only offered to a certain % of `stable` channel users until no serious show shopping widespread regressions arise)

There's no reason to guess or speculate though: everything is logged in a combination of the `data/updater.log` and standard Nextcloud log (e.g. `data/nextcloud.log`). So next time this happens - or if you happen to have some of your logs from when this last happened - those are what we need to look at to dig into this.

> The upgrade claims to succeed.

Do you recall the precise message? There are a couple different stopping points (probably not since this was a while I ago, but maybe it's come up again recently... or you have your logs from a prior upgrade where this occurred?)

How have your last few update/upgrade cycles been since this report?

**Comment by joshtrichards** [comment]
Duplicate of #401
